crates/ruff/src/rules/flake8_pyi/rules/duplicate_union_member.rs

```diff
@@ -0,0 +1,79 @@
+use rustc_hash::FxHashSet;
+use rustpython_parser::ast::{Expr, ExprKind, Operator};
+
+use ruff_diagnostics::{AlwaysAutofixableViolation, Diagnostic, Edit};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::comparable::ComparableExpr;
+use ruff_python_ast::helpers::unparse_expr;
+use ruff_python_ast::types::Range;
+
+use crate::checkers::ast::Checker;
+use crate::registry::AsRule;
+
+#[violation]
+pub struct DuplicateUnionMember {
+    pub duplicate_name: String,
+}
+
+impl AlwaysAutofixableViolation for DuplicateUnionMember {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Duplicate union member `{}`", self.duplicate_name)
+    }
+
+    fn autofix_title(&self) -> String {
+        format!("Remove duplicate union member `{}`", self.duplicate_name)
+    }
+}
+
+/// PYI016
+pub fn duplicate_union_member(checker: &mut Checker, expr: &Expr) {
+    let mut seen_nodes = FxHashSet::default();
+    traverse_union(&mut seen_nodes, checker, expr, None);
+}
+
+fn traverse_union<'a>(
+    seen_nodes: &mut FxHashSet<ComparableExpr<'a>>,
+    checker: &mut Checker,
+    expr: &'a Expr,
+    previous: Option<&'a Expr>,
+) {
+    // The union data structure usually looks like this:
+    //  a | b | c -> (a | b) | c
+    //
+    // However, parenthesized expressions can coerce it into any structure:
+    //  a | (b | c)
+    //
+    // So we have to traverse both branches in order (left, then right), to report duplicates
+    // in the order they appear in the source code.
+    if let ExprKind::BinOp {
+        op: Operator::BitOr,
+        left,
+        right,
+    } = &expr.node
+    {
+        // Traverse left subtree, then the right subtree, propagating the previous node.
+        traverse_union(seen_nodes, checker, left, previous);
+        traverse_union(seen_nodes, checker, right, Some(left));
+    }
+
+    // If we've already seen this union member, raise a violation.
+    if !seen_nodes.insert(expr.into()) {
+        let mut diagnostic = Diagnostic::new(
+            DuplicateUnionMember {
+                duplicate_name: unparse_expr(expr, checker.stylist),
```

Comment:
Changed this to use `unparse_expr`. E.g., if we have `str | str`, it'll now say `str` instead of `name`.